### PR TITLE
feat: OpenTelemetry support

### PR DESCRIPTION
**Pull request by gavin-aguiar**
## Description
Added support for opentelemetry.

The worker does not take a dependency on opentelemetry but checks if the function app has opentelemetry. If it finds the libraries, it will setup the Otel configs needed.

Otel is configured separately for sync and async. For async functions, we can directly call `configure_opentelemetry` but for sync functions this method has to be called inside the ThreadPool executor which is used to run sync functions. 
This is because OpenTelemetry is thread specific and will only update the Span on the thread it is configured

Sample function app.

```
import azure.functions as func 
import requests
import logging
from azure.monitor.opentelemetry import configure_azure_monitor

configure_azure_monitor()

app = func.FunctionApp()

@app.route('call_bing', auth_level=func.AuthLevel.ANONYMOUS)
def call_bing(req: func.HttpRequest) -> func.HttpResponse:
    logging.info(f'Logging to app insights...')
    res = requests.get('https://www.bing.com')
    return func.HttpResponse(f"Bing.com: status_code = {res.status_code}", status_code=res.status_code)
```

Fixes #

---

### PR information

- [ ] The title of the PR is clear and informative.
- [ ] There are a small number of commits, each of which has an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [ ] If applicable, the PR references the bug/issue that it fixes in the description.
- [ ] New Unit tests were added for the changes made and CI is passing.

### Quality of Code and Contribution Guidelines
- [ ] I have read the [contribution guidelines](https://github.com/Azure/azure-functions-python-worker/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: azure_functions_worker/dispatcher.py [hunk 1]
```diff
@@ -79,6 +80,11 @@ def __init__(self, loop: BaseEventLoop, host: str, port: int,
         self._function_metadata_result = None
         self._function_metadata_exception = None
 
+        # Used for checking if open telemetry is enabled
+        self.otel_libs_available = False
+        self.context_api = None
+        self.trace_context_propagator = None
```
Comment (YunchuWang) [inline azure_functions_worker/dispatcher.py]: nit: private vars? prefix with "_"?

File: azure_functions_worker/dispatcher.py [hunk 2]
```diff
@@ -262,6 +268,23 @@ async def _dispatch_grpc_request(self, request):
...
+                TraceContextTextMapPropagator)
+
+            self.context_api = context_api
+            self.trace_context_propagator = TraceContextTextMapPropagator()
+            self.otel_libs_available = True
+
+            logger.info("Successfully loaded OpenTelemetry modules. "
+                        "OpenTelemetry is now enabled.")
+        except ImportError:
+            self.otel_libs_available = False
```
Comment (YunchuWang) [inline azure_functions_worker/dispatcher.py]: log this error for troubleshooting?

File: azure_functions_worker/dispatcher.py [hunk 3]
```diff
@@ -291,6 +314,15 @@ async def _handle__worker_init_request(self, request):
...
         }
 
+        self.update_opentelemetry_status()
+
+        if self.otel_libs_available:
+            # When this capability is enabled, logs are not piped back to the
+            # host from the worker. Logs will directly go to where the user has
+            # configured them to go. This is to ensure that the logs are not
+            # duplicated.
+            capabilities["WorkerOpenTelemetryEnabled"] = _TRUE
```
Comment (YunchuWang) [inline azure_functions_worker/dispatcher.py]: nit: constant?

File: azure_functions_worker/dispatcher.py [hunk 4]
```diff
@@ -536,6 +568,9 @@ async def _handle__invocation_request(self, request):
                     args[name] = bindings.Out()
 
             if fi.is_async:
+                if self.otel_libs_available:
+                    self.configure_opentelemetry(fi_context)
```
Comment (YunchuWang) [inline azure_functions_worker/dispatcher.py:592]: if error raised here, is it caught for fail to invocation response?

File: azure_functions_worker/dispatcher.py [hunk 5]
```diff
@@ -744,6 +785,12 @@ async def _handle__close_shared_memory_resources_request(self, request):
                 request_id=self.request_id,
                 close_shared_memory_resources_response=response)
 
+    def configure_opentelemetry(self, invocation_context):
+        carrier = {TRACEPARENT: invocation_context.trace_context.trace_parent,
+                   TRACESTATE: invocation_context.trace_context.trace_state}
+        ctx = self.trace_context_propagator.extract(carrier)
+        self.context_api.attach(ctx)
```
Comment (YunchuWang) [inline azure_functions_worker/dispatcher.py]: how big is this ctx data, will this be a memory concern?

**Review by vrdmr — APPROVED** [review]

**Review by YunchuWang — APPROVED** [review]

**Pull request merged**